controllers/scopeinstance_controller.go

```diff
@@ -58,47 +67,166 @@ func (r *ScopeInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+	existingClusterRoleBindingList := &rbacv1.ClusterRoleBindingList{}
+
+	// Get the existing ClusterRoleBinding
+	if err := r.Client.List(ctx, existingClusterRoleBindingList, client.MatchingLabels{
+		"labels": "scopeInstanceController",
+	}); err != nil {
+		log.Log.Info("Error getting object", "error", err)
+	}
+
+	// look at namespaces - map of run time objects - think of this later
+	// Get the existing RoleBindings
+	if err := r.Client.List(ctx, existingRoleBindingList, client.MatchingLabels{
+		"labels": "scopeInstanceController",
+	}); err != nil {
+		log.Log.Info("Error getting object", "error", err)
+	}
+
+	//Create Map(Key, value) pair of existing Clusterrole binding.
+	// it will help to seperate the clusterrolebinding which we have to create and delete unnecessaty ones
+	clusterRoleBindingMap := make(map[string]string)
+	for _, csv := range existingClusterRoleBindingList.Items {
+		clusterRoleBindingMap[csv.RoleRef.Name] = csv.GetName()
+	}
+
+	//Create Map(Key, value) pair of existing role binding.
+	// it will help to seperate the rolebinding which we have to create and delete unnecessaty ones
+	roleBindingMap := make(map[string]string)
+	for _, csv := range existingRoleBindingList.Items {
+		roleBindingMap[csv.GetNamespace()+"-"+csv.RoleRef.Name] = csv.GetName()
+	}
+
 	objects := []client.Object{}
+	deleteObjects := []client.Object{}
+
+	// Check if any namespace is given in Scope Instance or not
 	if len(in.Spec.Namespaces) == 0 {
+
+		// it will create clusterrole as shown below if no namespace is provided
 		for _, cr := range st.Spec.ClusterRoles {
-			objects = append(objects, &rbacv1.ClusterRoleBinding{
-				ObjectMeta: metav1.ObjectMeta{
-					GenerateName: cr.GenerateName + "-",
-				},
-				Subjects: cr.Subjects,
-				RoleRef: rbacv1.RoleRef{
-					Kind: "ClusterRole",
-					Name: cr.GenerateName,
-				},
-			})
-		}
-	} else {
-		for _, namespace := range in.Spec.Namespaces {
-			for _, cr := range st.Spec.ClusterRoles {
-				objects = append(objects, &rbacv1.RoleBinding{
+			if val, ok := clusterRoleBindingMap[cr.GenerateName]; !ok {
+				objects = append(objects, &rbacv1.ClusterRoleBinding{
 					ObjectMeta: metav1.ObjectMeta{
 						GenerateName: cr.GenerateName + "-",
-						Namespace:    namespace,
+						Labels: map[string]string{
+							"labels": "scopeInstanceController",
+						},
+						// OwnerReferences: []metav1.OwnerReference{{
+						// 	APIVersion: in.APIVersion,
+						// 	Kind:       in.Kind,
+						// 	Name:       in.GetObjectMeta().GetName(),
+						// 	UID:        "d9607e19-f88f-11e6-a518-42010a800195",
+						// }},
 					},
 					Subjects: cr.Subjects,
 					RoleRef: rbacv1.RoleRef{
 						Kind: "ClusterRole",
 						Name: cr.GenerateName,
 					},
 				})
+
+			} else {
+				// delete from Map(key, value) if already created Cluster Role Bindings
+				delete(clusterRoleBindingMap, cr.GenerateName)
+				fmt.Println("Key found value is: ", val)
+			}
```

Comment:
We would need to do a comparison here before deleting the clusterRoleBinding from the list.

Reply:
yes,this will be take care in update process.